externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecutePython.scala

```diff
@@ -0,0 +1,238 @@
+import java.nio.file.{Files, Path, Paths}
+import java.util.concurrent.atomic.AtomicBoolean
+
+import scala.collection.JavaConverters._
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import org.apache.spark.api.python.KyuubiPythonGatewayServer
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types.StructType
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.operation.ArrayFetchIterator
+import org.apache.kyuubi.session.Session
+
+class ExecutePython(
+    session: Session,
+    override val statement: String,
+    worker: SessionPythonWorker) extends SparkOperation(session) {
+
+  override protected def resultSchema: StructType = {
+    if (result == null || result.schema.isEmpty) {
+      new StructType().add("output", "string")
+        .add("status", "string")
+        .add("ename", "string")
+        .add("evalue", "string")
+        .add("traceback", "array<string>")
+    } else {
+      result.schema
+    }
+  }
+
+  override protected def runInternal(): Unit = {
+    val response = worker.runCode(statement)
+    val output = response.map(_.content.getOutput()).getOrElse("")
+    val status = response.map(_.content.status).getOrElse("UNKNOWN_STATUS")
+    val ename = response.map(_.content.getEname()).getOrElse("")
+    val evalue = response.map(_.content.getEvalue()).getOrElse("")
+    val traceback = response.map(_.content.getTraceback()).getOrElse(Array.empty)
+    iter =
+      new ArrayFetchIterator[Row](Array(Row(output, status, ename, evalue, Row(traceback: _*))))
+  }
+
+}
+
+case class SessionPythonWorker(
+    errorReader: Thread,
+    pythonWorkerMonitor: Thread,
+    workerProcess: Process) {
+  private val stdin: PrintWriter = new PrintWriter(workerProcess.getOutputStream)
+  private val stdout: BufferedReader =
+    new BufferedReader(new InputStreamReader(workerProcess.getInputStream), 1)
+
+  def runCode(code: String): Option[PythonReponse] = {
+    val input = ExecutePython.toJson(Map("code" -> code, "cmd" -> "run_code"))
+    // scalastyle:off println
+    stdin.println(input)
+    // scalastyle:on
+    stdin.flush()
+    Option(stdout.readLine())
+      .map(ExecutePython.fromJson[PythonReponse](_))
+  }
+
+  def close(): Unit = {
+    val exitCmd = ExecutePython.toJson(Map("cmd" -> "exit_worker"))
+    // scalastyle:off println
+    stdin.println(exitCmd)
+    // scalastyle:on
+    stdin.flush()
+    stdin.close()
+    stdout.close()
+    errorReader.interrupt()
+    pythonWorkerMonitor.interrupt()
+    workerProcess.destroy()
+  }
+}
+
+object ExecutePython extends Logging {
+
+  // TODO:(fchen) get from conf
+  val pythonExec =
+    sys.env.getOrElse("PYSPARK_PYTHON", sys.env.getOrElse("PYSPARK_DRIVER_PYTHON", "python3"))
+  private val isPythonGatewayStart = new AtomicBoolean(false)
+  val kyuubiPythonPath = Files.createTempDirectory("")
+  def init(): Unit = {
+    if (!isPythonGatewayStart.get()) {
+      synchronized {
```

Comment:
Is it thread safe here? it's locking the code block itself rather than syncing the AtomicBoolean and with no double check.

Reply:
added double-check.